File: scripts/NumberPublisher.py
#!/usr/bin/python
import rospy
from openhab_msgs.msg import NumberCommand

class NumberPublisher(object):
    """Node example class."""

    def __init__(self, item_name):

        self.item_name = item_name
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", NumberCommand, queue_size=10)
        self.rate = rospy.Rate(10) # 10hz

        # Initialize message variables.
        self.enable = False
        self.message = ""

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        """Turn on publisher."""
        self.enable = True
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", NumberCommand, queue_size=10)

        i = 0
        while not rospy.is_shutdown():

            self.message = NumberCommand()

            i = i + 1
            self.message.command = float(i)
            
            self.message.isnull = False

            self.message.header.stamp = rospy.Time.now()
            self.message.header.frame_id = "/base_link"
            self.message.item = self.item_name

            message = "Publishing to %s at %s: command = %s" % (self.message.item, rospy.get_time(), self.message.command)
            rospy.loginfo(message)

            self.pub.publish(self.message)
            self.rate.sleep()

    def stop(self):
        """Turn off publisher."""
        self.enable = False
        self.pub.unregister()

# Main function.
if __name__ == "__main__":
    # Initialize the node and name it.
    rospy.init_node("NumberPublisherNode", anonymous=True)
    # Go to class functions that do all the heavy lifting.

    numberPublisher = NumberPublisher("testNumber")

    try:
        numberPublisher.start()
    except rospy.ROSInterruptException:
        pass
    # Allow ROS to go to all callbacks.
    # spin() simply keeps python from exiting until this node is stopped
    rospy.spin()

File: src/NumberPublisher.cpp
#include "ros/ros.h"
#include "openhab_msgs/NumberCommand.h"

class NumberPublisher {
public:
  NumberPublisher(std::string item_name) : item_name_(item_name), enable_(false) {
    pub_ = nh_.advertise<openhab_msgs::NumberCommand>("/openhab/items/" + item_name_ + "/command", 10);
    rate_ = ros::Rate(10);
    if (enable_) {
      start();
    } else {
      stop();
    }
  }

  void start() {
    enable_ = true;
    pub_ = nh_.advertise<openhab_msgs::NumberCommand>("/openhab/items/" + item_name_ + "/command", 10);
    int i = 0;
    while (ros::ok()) {
      openhab_msgs::NumberCommand msg;
      i++;
      msg.command = static_cast<float>(i);
      msg.isnull = false;
      msg.header.stamp = ros::Time::now();
      msg.header.frame_id = "/base_link";
      msg.item = item_name_;
      std::string message = "Publishing to " + msg.item + " at " + std::to_string(ros::Time::now().toSec()) +
                            ": command = " + std::to_string(msg.command);
      ROS_INFO("%s", message.c_str());
      pub_.publish(msg);
      rate_.sleep();
    }
  }

  void stop() {
    enable_ = false;
    pub_.shutdown();
  }

private:
  ros::NodeHandle nh_;
  ros::Publisher pub_;
  ros::Rate rate_;
  std::string item_name_;
  bool enable_;
};

int main(int argc, char **argv) {
  // Initialize the node and name it.
  ros::init(argc, argv, "NumberPublisherNode");
  NumberPublisher numberPublisher("testNumber");
  try {
    numberPublisher.start();
  } catch (const ros::Exception &e) {
    // Do nothing
  }
  // spin() simply keeps the node running until it's stopped
  ros::spin();
  return 0;
}
